File: taller-threejs/threejs/src/Pokebola.js
/**
 Pokebola
 Clase pokebola.
 @author Pablo Pizarro R. @ppizarror.com
 @date 2018-2019
 @license MIT
 */
"use strict";

/**
 * Crea una Pokebola.
 *
 * @param {Point2D} pos_inicial - Posición inicial
 * @param {Point2D} velocidad_inicial - Velocidad inicial
 * @param {number} diametro_pokebola - Diámetro pokebola
 * @param {number} color_pokebola - Color pokebola
 * @param {number} alpha_elastico - Coeficiente de rebote
 * @param {number} plano - Posición del plano
 * @param {number} plano_largo - Largo del plano
 * @constructor
 */
function Pokebola(pos_inicial, velocidad_inicial, diametro_pokebola, color_pokebola,
                  alpha_elastico, plano, plano_largo) {

    // noinspection ES6ConvertVarToLetConst
    /**
     * Puntero al objeto
     */
    var self = this;

    /**
     * Guarda las variables
     */
    this._alpha = alpha_elastico;
    this._color = color_pokebola;
    this._dt = 1 / 60; // Incremento de tiempo
    this._g = 98.1665; // "Gravedad"
    this._pos_plano = plano;
    this._pos = pos_inicial;
    this._plano_largo = plano_largo;
    this._radio = diametro_pokebola / 2;
    this._vel = velocidad_inicial;

    /**
     * Crea el modelo en three.js
     */
    this.crear_modelo = function (scene) {

        /**
         * Crea una esfera
         * Basado en: https://threejs.org/docs/#api/en/geometries/SphereGeometry
         */
        let geometry = new THREE.SphereGeometry(this._radio, 32, 32);
        let material = new THREE.MeshPhongMaterial({color: this._color, dithering: true});
        this._sphere = new THREE.Mesh(geometry, material);
        this._sphere.castShadow = true;
        scene.add(this._sphere);
    };

    /**
     * Actualizar posición
     */
    this.actualizar_posicion = function () {

        // Actualiza velocidad en z (gravedad)
        self._vel.set_z(self._vel.get_z() - self._g * self._dt);  // v(t) = v(t-1) + a*dt

        // Actualiza la posición
        self._pos.set_x(self._pos.get_x() + self._vel.get_x() * self._dt);
        self._pos.set_y(self._pos.get_y() + self._vel.get_y() * self._dt);
        self._pos.set_z(self._pos.get_z() + self._vel.get_z() * self._dt);

        // Busca colisiones
        self._detectar_colision();

        // Ubica el modelo
        self._mover_modelo();

    };

    /**
     * Detecta las colisiones.
     */
    this._detectar_colision = function () {
        if (self._pos.get_z() - self._radio < self._pos_plano) { // Colisión inferior
            self._vel.set_z(-self._vel.get_z() * self._alpha);
            self._pos.set_z(self._pos_plano + self._radio);
        }
        if (self._pos.get_x() + self._radio > self._plano_largo) { // Choque en x +
            self._vel.set_x(-self._vel.get_x() * self._alpha);
            self._pos.set_x(self._plano_largo - self._radio);
        }
        if (self._pos.get_x() - self._radio < -self._plano_largo) { // Choque en x -
            self._vel.set_x(-self._vel.get_x() * self._alpha);
            self._pos.set_x(-self._plano_largo + self._radio);
        }
        if (self._pos.get_y() + self._radio > self._plano_largo) { // Choque en y +
            self._vel.set_y(-self._vel.get_y() * self._alpha);
            self._pos.set_y(self._plano_largo - self._radio);
        }
        if (self._pos.get_y() - self._radio < -self._plano_largo) { // Choque en y -
            self._vel.set_y(-self._vel.get_y() * self._alpha);
            self._pos.set_y(-self._plano_largo + self._radio);
        }
    };

    /**
     * Mueve el modelo.
     */
    this._mover_modelo = function () {
        this._sphere.position.x = this._pos.get_x();
        this._sphere.position.y = this._pos.get_z();
        this._sphere.position.z = this._pos.get_y();
    };

}

File: taller-opcional/taller-threejs/python/pokebola.py
"""
Clase pokebola. La idea de esta clase es definir todos los
métodos propios de este objeto:
    - Detectar colisión con el plano
    - Modificar su velocidad
    - Dibujarse

@author Pablo Pizarro R. @ppizarror
@date 2018-2019
@license MIT
"""

# Importación de librerías
from PyOpenGLtoolbox import *


class Pokebola:
    def __init__(self, pos_inicial, velocidad_inicial, diametro_pokebola, color_pokebola,
                 alpha_elastico, plano):
        """
        Constructor

        @param pos_inicial: Posición inicial de la Pokebola
        @param velocidad_inicial: Velocidad inicial de la Pokebola
        @param diametro_pokebola: Radio de la Pokebola
        @param color_pokebola: Color de la Pokebola
        @param alpha_elastico: Constante elática de rebote
        @param plano: Posición en z del plano
        @type pos_inicial: Point3
        @type velocidad_inicial: Vector3
        @type diametro_pokebola: float
        @type color_pokebola: list
        @type alpha_elastico: float
        @type plano: float
        """
        self._alpha = alpha_elastico
        self._color = color_pokebola
        self._dt = 1 / 60  # Incremento de tiempo, idealmente se debería calcular y no asumir constante
        self._g = 9.8  # "Gravedad" del mundo
        self._modelo = load_gmsh_model("esfera.msh", diametro_pokebola, 0, 0, 0, True, False)
        self._pos = pos_inicial
        self._pos_plano = plano
        self._radio = diametro_pokebola / 2
        self._vel = velocidad_inicial

    def actualizar_posicion(self):
        """
        La idea de esta función es actualizar la posición de la pokebola, para ello
        se usa un contador interno del tiempo (que se incrementa cada vez que se llama
        a esta función). Luego se aplican las funciones matemáticas para modificar la velocidad
        (intro a la física).

        Si se choca con el plano entonces se rebota. El plano se define entre (-50,-50) y (50,50),
        si las coordenadas (x,y) están dentro de ese cuadrante y la posición inferior de la pokebola
        (o sea, centro-radio) es igual o inferior al plano rebota
        :return:
        """

        # Actualiza velocidad en z (gravedad)
        self._vel.set_z(self._vel.get_z() - self._g * self._dt)  # v(t) = v(t-1) + a*dt

        # Actualiza la posición
        self._pos.set_x(self._pos.get_x() + self._vel.get_x() * self._dt)
        self._pos.set_y(self._pos.get_y() + self._vel.get_y() * self._dt)
        self._pos.set_z(self._pos.get_z() + self._vel.get_z() * self._dt)

        # Busca colisiones
        self._detectar_colision()

    def _detectar_colision(self):
        """
        Maneja las colisiones
        :return:
        """

        # Colisión con el plano
        if self._pos.get_z() - self._radio <= self._pos_plano:
            self._vel.set_z(-self._vel.get_z() * self._alpha)
            self._pos.set_z(self._pos_plano + self._radio)

    def dibujar(self):
        """
        Dibuja el modelo
        :return:
        """
        self._modelo.draw(self._pos.export_to_list(), self._color)
